Add Workstation Session Control

chrisvanrun opened the pull request:
Part of the pitch:
- https://github.com/DIAGNijmegen/rse-roadmap/issues/186

Selective copy of relevant base code from private

jmsmkn reviewed (changes requested):
On grand-challenge.org sessions can be in one of many regions. `us-east-1` is hardcoded in all of the data-session-origin. 

Should we handle workstations that do not respond to the request? At the moment it will look like all the buttons are broken after the first one and will require the user to manually close the tab.

jmsmkn commented on app/grandchallenge/workstations/static/workstations/js/session-control.js:
Needs to be handled.

amickan replied:
Will be done in seperate PR, I suggest

jmsmkn commented on app/grandchallenge/algorithms/templates/algorithms/job_list_row.html:
```diff
@@ -66,11 +67,15 @@
+            title="View algorithm result"
+            data-session-control
+            data-session-origin="https://us-east-1.{{ request.site.domain }}"
+            data-create-session-url='{% url 'workstations:workstation-session-create' slug=object.algorithm_image.algorithm.workstation.slug %}'
+            data-workstation-query='{% workstation_query algorithm_job=object config=object.algorithm_image.algorithm.workstation_config %}'
```
The lookup for the slug and config has changed here compared to the LHS. Does this introduce new n+1 queries?

amickan replied:
It doesn't make a difference because we're prefetching the algorithm on the jobs in the view, but I will anyway change this back to the more direct lookup.

jmsmkn commented on app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_progress.html:
```diff
@@ -52,10 +53,14 @@ <h2>Users Progress for {{ reader_study.title }}</h2>
+                                    data-create-session-url='{% url 'workstations:workstation-session-create' slug=reader_study.workstation.slug %}'
+                                    data-session-origin="https://us-east-1.{{ request.site.domain }}"
+                                    data-workstation-query='{% workstation_query reader_study=reader_study user=user.obj %}'
+                                    data-workstation-window-identifier='workstation-{{ object|meta_attr:'app_label' }}'>
```
Should this be `object` or `reader_study`?

jmsmkn commented on app/grandchallenge/reader_studies/templates/reader_studies/readerstudy_images_row.html:
```diff
@@ -12,11 +13,14 @@
+   data-create-session-url='{% url 'workstations:workstation-session-create' slug=reader_study.workstation.slug %}'
+   data-session-origin="https://us-east-1.{{ request.site.domain }}"
+   data-workstation-query='{% workstation_query image=object config=reader_study.workstation_config %}'
+   data-workstation-window-identifier='workstation-{{ object|meta_attr:'app_label' }}'
```
Should this be `object` (an image) or `reader_study`?

jmsmkn commented on app/grandchallenge/archives/templates/archives/archive_items_row.html:
```diff
@@ -26,8 +27,12 @@
+   data-create-session-url='{% url 'workstations:workstation-session-create' slug=archive.workstation.slug %}'
+   data-session-origin="https://us-east-1.{{ request.site.domain }}"
+   data-workstation-query='{% workstation_query archive_item=object config=archive.workstation_config %}'
+   data-workstation-window-identifier='workstation-{{ object|meta_attr:'app_label' }}-{{ object|meta_attr:'model_name' }}'>
```
Should these be `object` (archive item) or `archive`?

jmsmkn commented on app/grandchallenge/archives/templates/archives/archive_cases_row.html:
```diff
@@ -42,11 +43,14 @@
+        data-create-session-url='{% url 'workstations:workstation-session-create' slug=archive.workstation.slug %}'
+        data-session-origin="https://us-east-1.{{ request.site.domain }}"
+        data-workstation-query='{% workstation_query image=object config=archive.workstation_config %}'
+        data-workstation-window-identifier='workstation-{{ object|meta_attr:'app_label' }}-{{ object|meta_attr:'model_name' }}'>
```
Should these be `object` (image) or `archive`? Is `model_name` necessary?

amickan replied:
I would say `archive`

jmsmkn commented on app/grandchallenge/algorithms/templates/algorithms/job_list_row.html:
```diff
@@ -66,11 +67,15 @@
+            data-session-origin="https://us-east-1.{{ request.site.domain }}"
+            data-create-session-url='{% url 'workstations:workstation-session-create' slug=object.algorithm_image.algorithm.workstation.slug %}'
+            data-workstation-query='{% workstation_query algorithm_job=object config=object.algorithm_image.algorithm.workstation_config %}'
+            data-workstation-window-identifier='workstation-{{ object|meta_attr:'app_label' }}-{{ object|meta_attr:'model_name' }}'>
```
Is model_name necessary here?

amickan replied:
From our discussions with Chris yesterday, I would we say we don't need the model name. All we want to be able to distinguish is reader_study sessions from algorithm from archive sessions. The app_label will do for that.

amickan commented:
@jmsmkn we were planning on addressing the issues you mention above (as well as other improvements) in future PRs to this feature branch, and just move over the existing code from private to the feature branch as a start. That's why it's a blind copy of the code. On second thought, we might oversee problems that way, so I'll make changes to this PR to address the core issues.

jmsmkn commented:
Ah OK, missed the base branch.

amickan commented:
I have updated the window-identifiers to no longer include the `model_name` and have updated the reference objects where necessary. All other changes, i.e. checking for active sessions, retrieving the session region and handling unresponsive workstation sessions will be added in future PRs.

chrisvanrun commented:
> @jmsmkn we were planning on addressing the issues you mention above (as well as other improvements) in future PRs to this feature branch, and just move over the existing code from private to the feature branch as a start. That's why it's a blind copy of the code. On second thought, we might oversee problems that way, so I'll make changes to this PR to address the core issues.

I agree with the possibility of oversights. I vaguely recall a buried reason for using the model name as well for one of the apps, but we'll run into that eventually when working on this.
